spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java

```diff
@@ -0,0 +1,312 @@
+
+	/**
+	 * Sets the {@link Duration} to wait for a file to be loaded into BigQuery before timing out
+	 * when waiting synchronously.
+	 * @param timeout the {@link Duration} timeout to wait for a file to load
+	 */
+	public void setTimeout(Duration timeout) {
+		Assert.notNull(timeout, "Timeout duration must not be null.");
+		this.timeout = timeout;
+	}
+
+	/**
+	 * Sets the {@link Duration} amount of time to wait between successive polls on the status of
+	 * a BigQuery job.
+	 * @param pollInterval the {@link Duration} poll interval for BigQuery job status polling
+	 */
+	public void setJobPollInterval(Duration pollInterval) {
+		Assert.notNull(pollInterval, "Poll interval duration must not be null.");
+		this.jobPollInterval = pollInterval;
+	}
+
+	/**
+	 * A {@code boolean} indicating if the {@link BigQueryFileMessageHandler} should synchronously
+	 * wait for each file to be successfully loaded to BigQuery.
+	 * @param sync whether {@link BigQueryFileMessageHandler} should wait synchronously for jobs to
+	 * 		complete
+	 */
+	public void setSync(boolean sync) {
+		this.sync = sync;
+	}
+
+	@Override
+	protected Object handleRequestMessage(Message<?> message) {
+		String datasetName =
+				this.datasetNameExpression.getValue(this.evaluationContext, message, String.class);
+		String tableName =
+				this.tableNameExpression.getValue(this.evaluationContext, message, String.class);
+		FormatOptions formatOptions =
+				this.formatOptionsExpression.getValue(this.evaluationContext, message, FormatOptions.class);
+
+		Assert.notNull(datasetName, "BigQuery dataset name must not be null.");
+		Assert.notNull(tableName, "BigQuery table name must not be null.");
+		Assert.notNull(formatOptions, "Data file formatOptions must not be null.");
+
+		TableId tableId = TableId.of(datasetName, tableName);
+
+		WriteChannelConfiguration writeChannelConfiguration = WriteChannelConfiguration
+				.newBuilder(tableId)
+				.setFormatOptions(formatOptions)
+				.setAutodetect(this.autoDetectSchema)
+				.setWriteDisposition(this.writeDisposition)
+				.build();
+
+		TableDataWriteChannel writer = bigQuery.writer(writeChannelConfiguration);
+
+		try (
+				InputStream inputStream = convertToInputStream(message.getPayload());
+				OutputStream sink = Channels.newOutputStream(writer)) {
+			// Write data from data input file to BigQuery
+			StreamUtils.copy(inputStream, sink);
+		}
+		catch (FileNotFoundException e) {
+			throw new MessageHandlingException(
+					message, "Failed to find file to write to BigQuery in message handler: " + this, e);
+		}
+		catch (IOException e) {
+			throw new MessageHandlingException(
+					message, "Failed to write data to BigQuery tables in message handler: " + this, e);
+		}
+
+		if (writer.getJob() == null) {
+			throw new MessageHandlingException(
+					message, "Failed to initialize the BigQuery write job in message handler: " + this);
+		}
+
+		// Prepare the polling task for the ListenableFuture result returned to end-user
```

Comment:
I'd take lines 255 through 275 into a helper method "watchProgress" or "watchUploadStatus" or something that returns `SettableListenableFuture<Job> result`. `handleListenableMessage()` is getting really long.